### ISSUE TITLE
Missing "Condition will always evaluate to False since the types "UUID" and "list[UUID]" have no overlap"

### ISSUE THREAD

**Issue by Etienne-Leveque** [microsoft/pyright#7354]
**Describe the bug**
When I compare a list of UUIDs with an UUID, I don't get unnecessary comparison report.

**Code or Screenshots**

```python
from uuid import UUID, uuid4

a: UUID = uuid4()
b: list[UUID] = [uuid4()]

if a == b: # No reportUnnecessaryComparison
    ...

class Bar:
    ...

c: Bar = Bar()
d: list[Bar] = [Bar()]

if c == d: # Condition will always evaluate to False since the types "Bar" and "list[Bar]" have no overlap
    ...
```

**VS Code extension or command-line**
pyright CLI 1.1.351

**Comment by erictraut** [comment]
This behavior is by design. In Python, objects can override `__eq__` and `__ne__` in any manner they choose. Sometimes it is intentional that they support comparisons without overlap. For that reason, the `reportUnnecessaryComparison` diagnostic is suppressed if their operand type implements a custom `__eq__` or `__ne__` method. In your example above, `Bar` does not implement such a method, but `UUID` does.

**Comment by Etienne-Leveque** [comment]
Thanks for the explanation and the quick reply!

**Comment by Holt59** [comment]
@erictraut I have a similar issue when using `in` with built-in types, e.g.

```python
def f(xu: dict[tuple[int, str], str], u: int):
    assert u in xu  # <- no error
```

Does this also fall in the by design as explained above?

**Comment by erictraut** [comment]
@Holt59, this check is suppressed for expressions in an `assert` statement. The theory is that if you're adding a runtime assertion, you probably know something more than the type checker. If you move the `u in xu` expression outside of an assert (e.g. in an `if` statement), you'll see that the `reportUnnecessaryContains` error is generated.
